opentelemetry-api/src/opentelemetry/attributes/__init__.py

```diff
@@ -118,6 +118,98 @@ def _clean_attribute(
     return None
 
 
+def _clean_extended_attribute_value(
+    value: types.AnyValue, max_len: Optional[int]
+) -> types.AnyValue:
+    # for primitive types just return the value and eventually shorten the string length
+    if value is None or isinstance(value, _VALID_ATTR_VALUE_TYPES):
+        if max_len is not None and isinstance(value, str):
+            value = value[:max_len]
+        return value
+
+    if isinstance(value, Mapping):
+        cleaned_dict: dict[str, types.AnyValue] = {}
+        for key, element in value.items():
+            # skip invalid keys
+            if not (key and isinstance(key, str)):
+                _logger.warning(
+                    "invalid key `%s`. must be non-empty string.", key
+                )
+                continue
+
+            cleaned_dict[key] = _clean_extended_attribute(
+                key=key, value=element, max_len=max_len
+            )
+
+        return cleaned_dict
+
+    if isinstance(value, Sequence):
+        sequence_first_valid_type = None
+        cleaned_seq: list[types.AnyValue] = []
+
+        for element in value:
+            if element is None:
+                cleaned_seq.append(element)
+                continue
+
+            if max_len is not None and isinstance(element, str):
+                element = element[:max_len]
+
+            element_type = type(element)
+            if element_type not in _VALID_ATTR_VALUE_TYPES:
+                element = _clean_extended_attribute_value(
+                    element, max_len=max_len
+                )
+                element_type = type(element)  # type: ignore
+
+            # The type of the sequence must be homogeneous. The first non-None
+            # element determines the type of the sequence
+            if sequence_first_valid_type is None:
+                sequence_first_valid_type = element_type
+            # use equality instead of isinstance as isinstance(True, int) evaluates to True
+            elif element_type != sequence_first_valid_type:
+                _logger.warning(
+                    "Mixed types %s and %s in attribute value sequence",
+                    sequence_first_valid_type.__name__,
+                    type(element).__name__,
+                )
+                return None
```

Comment:
It looks like we drop the whole list if there is mixed type vs just dropping the bad values. Is this specified anywhere?

Reply:
Asked on otel-specification and they have said dropping everything is fine